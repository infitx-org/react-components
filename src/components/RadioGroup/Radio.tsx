import React, { ChangeEvent } from "react";
import classnames from "classnames";
import { Kind, OptionValue } from "../../types";

export type RadioProps = {
  kind?: `${Kind}`;
  id?: string;
  name?: string;
  checked: boolean;
  label?: string;
  value: OptionValue;
  disabled?: boolean;
  vertical?: boolean;
  onChange: (e: ChangeEvent<HTMLInputElement>) => void;
};

function Radio({
  kind = Kind.Primary,
  id,
  name,
  onChange,
  checked,
  value,
  label,
  disabled = false,
  vertical = false,
}: RadioProps) {
  const inputClassName = classnames([
    "rc-radio__input",
    `rc-radio__input--${kind}`,
  ]);
  const wrapperClassName = classnames([
    "rc-radio__wrapper",
    vertical && "rc-radio__wrapper--vertical",
  ]);
  const labelClassName = classnames([
    "rc-radio__label",
    disabled && "rc-radio__label--disabled",
  ]);
  return (
    <label className={wrapperClassName}>
      <input
        id={id}
        type="radio"
        name={name}
        className={inputClassName}
        onChange={onChange}
        checked={checked}
        value={value.toString()}
        disabled={disabled}
      />
      <span className={labelClassName}>{label}</span>
    </label>
  );
}

export default React.memo(Radio);
