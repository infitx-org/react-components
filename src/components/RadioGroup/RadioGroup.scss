@use "sass:map";
@use "../../scss/base/shadows.scss" as *;
@use "../../scss/base/colors.scss" as *;
@use "../../scss/themes.scss" as *;
@use "../../scss/mixins.scss" as *;


.rc-radiogroup {
  @include rc-component;
  padding: 3px;
  font-size: 14px;
  .rc-radiogroup__inputs {
    display: flex;
    align-items: center;
  }
}

.rc-radiogroup--vertical {
  .rc-radiogroup__inputs {
    flex-direction: column;
    align-items: flex-start;
  }
}

.rc-radiogroup__label {
  display: inline-block;
  margin-bottom: 5px;
}

.rc-radio__wrapper {
  display: inline-flex;
  margin-right: 10px;
  align-items: center;
}

.rc-radio__wrapper--vertical {
  margin-bottom: 6px;
}

.rc-radio__label {
  cursor: pointer;
  display: inline-block;
}

.rc-radio__label--disabled {
  color: #999;
  cursor: default;
}

.rc-radio__input {
  appearance: none;
  width: 20px;
  height: 20px;
  margin: 0;
  margin-right: 5px;
  display: inline-block;
  outline: none;
  position: relative;
  box-shadow: 0px 0px 0px 0px rgba(0, 0, 0, 0.13);
  border: 1px solid #cacbcc;
  border-radius: 50%;
  transition: all 0.2s ease-out;

  &:hover {
    box-shadow: $shadow-size rgba(0, 0, 0, 0.13);
  }

  &:not(:disabled) {
    cursor: pointer;
  }

  &:disabled {
    background: #eee;
    box-shadow: none;
    &:checked {
      border-color: #60d2fd;
      background: #60d2fd;
    }
  }

  &:checked {
    &:after {
      content: "";
      box-sizing: content-box;
      box-sizing: initial;
      position: absolute;
      left: 5px;
      top: 5px;
      border-radius: 50%;
      background: #fff;
      box-shadow: 0px 1px 2px rgba(0, 0, 0, 0.2);
      width: 8px;
      height: 8px;
      animation: animateRadioContent 0.2s ease-out 1;
    }
  }
}

@each $kind in $kinds {
  .rc-radio__input--#{$kind} {
    &:checked {
      border-color: map.get($theme-kinds--dark, $kind);
      background: linear-gradient(
        0deg,
        map.get($theme-kinds, $kind),
        map.get($theme-kinds--dark, $kind)
      );
    }
    &:checked:disabled {
      border-color: map.get($theme-kinds--light, $kind);
      background: map.get($theme-kinds--light, $kind);
    }
    &:focus {
      box-shadow: $shadow-size map.get($theme-kinds-shadows, $kind);
    }
  }
}

@keyframes animateRadioContent {
  0% {
    opacity: 0;
    transform: scale(0.5, 0.5);
  }
  100% {
    opacity: 1;
    transform: scale(1, 1);
  }
}
