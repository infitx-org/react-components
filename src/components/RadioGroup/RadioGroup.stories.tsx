/* eslint no-console: "off" */
import React, { useState } from "react";

import RadioGroup from "./RadioGroup";

export default {
  title: "Components/RadioGroup",
  component: RadioGroup,
};

function optionMaker(item, index) {
  return {
    label: `option number ${index + 1}`,
    value: (index + 1).toString(),
  };
}

const options = [
  {
    label: "disabled",
    value: "disabled",
    disabled: true,
  },
  ...[...Array(3)].map(optionMaker),
];

export const Default = () => {
  const [selected, setSelected] = useState("1");
  return (
    <RadioGroup
      label="Default Label"
      options={options}
      onChange={(e) => setSelected(e.target.value)}
      selected={selected}
    />
  );
};

export const AllDisabled = () => {
  const [selected, setSelected] = useState("1");
  return (
    <RadioGroup
      label="All Disabled Label"
      options={options}
      onChange={(e) => setSelected(e.target.value)}
      selected={selected}
      disabled
    />
  );
};

export const SelectedDisabled = () => {
  const [selected, setSelected] = useState("disabled");
  return (
    <RadioGroup
      label="Selected Disabled Label"
      options={options}
      onChange={(e) => setSelected(e.target.value)}
      selected={selected}
    />
  );
};

export const VerticalAlignment = () => {
  const [selected, setSelected] = useState("1");
  return (
    <RadioGroup
      label="Vertical Alignment Label"
      options={options}
      onChange={(e) => setSelected(e.target.value)}
      vertical
    />
  );
};

export const Variants = () => {
  const [selected, setSelected] = useState("disabled");
  const c = (
    <RadioGroup
      label="Vertical Alignment Label"
      options={options}
      selected={selected}
      onChange={(e) => setSelected(e.target.value)}
    />
  );
  return (
    <>
      {[
        "primary",
        "secondary",
        "tertiary",
        "success",
        "danger",
        "warning",
        "dark",
        "light",
      ].map((kind) =>
        React.cloneElement(c, { ...c.props, kind, label: kind, name: kind })
      )}
    </>
  );
};
