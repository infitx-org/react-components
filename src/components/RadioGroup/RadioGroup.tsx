import React, { ChangeEvent } from "react";
import classnames from "classnames";
import Radio, { RadioProps } from "./Radio";
import "./RadioGroup.scss";

type Option = Pick<RadioProps, "label" | "value" | "disabled">;

interface RadioGroupProps {
  id?: string;
  name: string;
  label: string;
  selected?: string;
  disabled?: boolean;
  options: Option[];
  vertical?: boolean;
  onChange: (e: ChangeEvent<HTMLInputElement>) => void;
}

function RadioGroup({
  id,
  name,
  label,
  disabled,
  selected,
  options,
  vertical,
  onChange,
}: RadioGroupProps) {
  const [selectedOption, setSelected] = React.useState(selected);
  React.useEffect(() => {
    setSelected(selected);
  }, [selected]);
  const compositeOnChange = (e: ChangeEvent<HTMLInputElement>) => {
    setSelected(e.target.value);
    onChange?.(e);
  };
  const classNames = classnames([
    "mb-input",
    "input-radiogroup",
    vertical && "input-radiogroup--vertical",
  ]);
  return (
    <div className={classNames} id={id}>
      {label && <span>{label}</span>}
      <div className="input-radiogroup__inputs" id={id}>
        {options.map((option: Option, index: number) => (
          <Radio
            key={index.toString()}
            id={`${name}-${index}`}
            onChange={compositeOnChange}
            name={name}
            checked={selectedOption === option.value}
            label={option.label}
            value={option.value}
            disabled={option.disabled || disabled}
          />
        ))}
      </div>
    </div>
  );
}

export default React.memo(RadioGroup);
