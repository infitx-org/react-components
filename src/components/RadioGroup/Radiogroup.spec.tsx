import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import userEvent from "@testing-library/user-event";
import RadioGroup from "./RadioGroup";

const { log } = console;

const options = new Array(5)
  .fill(0)
  .map((_, index) => ({ label: index.toString(), value: index.toString() }));

const commonProps = {
  options,
  name: "test-name",
  onChange: jest.fn(),
  label: "test-label",
};

describe("tests the radio group", () => {
  it("renders the option group", () => {
    const { container } = render(<RadioGroup {...commonProps} />);
    expect(container.querySelector(".input-radiogroup span")).toBeTruthy();
    expect(container.querySelectorAll('input[type="radio"]')).toHaveLength(5);
  });

  it("renders the label prop", () => {
    const { container } = render(<RadioGroup {...commonProps} />);
    expect(container.querySelector(".input-radiogroup span")).toHaveTextContent(
      "test-label"
    );
  });

  it("renders the id prop", () => {
    const { container } = render(<RadioGroup {...commonProps} id="test-id" />);
    expect(container.querySelector(".input-radiogroup")).toHaveAttribute(
      "id",
      "test-id"
    );
  });

  it("renders the name prop", () => {
    const { container } = render(<RadioGroup {...commonProps} />);
    const inputs = container.querySelectorAll('input[type="radio"]');
    inputs.forEach((input) => {
      expect(input).toHaveAttribute("name", "test-name");
    });
  });

  it("renders selected prop", () => {
    const { container } = render(<RadioGroup {...commonProps} selected="1" />);
    const inputs = container.querySelectorAll('input[type="radio"][checked]');
    expect(inputs).toHaveLength(1);
  });

  it("renders disabled prop", () => {
    const { container } = render(<RadioGroup {...commonProps} disabled />);
    const inputs = container.querySelectorAll('input[type="radio"]');
    inputs.forEach((input) => {
      expect(input).toBeDisabled();
    });
  });

  it("triggers the onChange", () => {
    const mockEvent = jest.fn();
    const { container } = render(
      <RadioGroup {...commonProps} onChange={mockEvent} />
    );
    userEvent.click(container.querySelector('input[type="radio"][value="1"]'));
    expect(mockEvent).toHaveBeenCalledWith(
      expect.objectContaining({
        target: expect.objectContaining({ value: "1" }),
      })
    );
  });
});

// Snapshot testing
it("renders the Icon correctly when multiple props are set", () => {
  const { container } = render(<RadioGroup {...commonProps} />);
  expect(container).toMatchSnapshot();
});
