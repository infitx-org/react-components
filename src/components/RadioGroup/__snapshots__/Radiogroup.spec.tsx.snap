// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the Icon correctly when multiple props are set 1`] = `
<div>
  <div
    class="mb-input input-radiogroup"
  >
    <span
      class="input-radiogroup__label"
    >
      test-label
    </span>
    <div
      class="input-radiogroup__inputs"
    >
      <label
        class="input-radio__wrapper"
      >
        <input
          class="input-radio__input input-radio__input--primary"
          name="test-name"
          type="radio"
          value="0"
        />
        <span
          class="input-radio__label"
        >
          0
        </span>
      </label>
      <label
        class="input-radio__wrapper"
      >
        <input
          class="input-radio__input input-radio__input--primary"
          name="test-name"
          type="radio"
          value="1"
        />
        <span
          class="input-radio__label"
        >
          1
        </span>
      </label>
      <label
        class="input-radio__wrapper"
      >
        <input
          class="input-radio__input input-radio__input--primary"
          name="test-name"
          type="radio"
          value="2"
        />
        <span
          class="input-radio__label"
        >
          2
        </span>
      </label>
      <label
        class="input-radio__wrapper"
      >
        <input
          class="input-radio__input input-radio__input--primary"
          name="test-name"
          type="radio"
          value="3"
        />
        <span
          class="input-radio__label"
        >
          3
        </span>
      </label>
      <label
        class="input-radio__wrapper"
      >
        <input
          class="input-radio__input input-radio__input--primary"
          name="test-name"
          type="radio"
          value="4"
        />
        <span
          class="input-radio__label"
        >
          4
        </span>
      </label>
    </div>
  </div>
</div>
`;
